bpo-43352: Add a Barrier object in asyncio lib (#24903)
python/cpython: Doc/library/asyncio-sync.rst

   a :exc:`ValueError` in :meth:`~Semaphore.release` if it
   increases the internal counter above the initial *value*.

Barrier
=======

> .. class:: Barrier(parties, action=None)

Reviewer: Please, could you explain a little more on documentation? I mean adding some methods docs, and examples?

Author: Ok I am working on today

Author: Update documentation with a sample and a complete description of the class behavior

   certain number of tasks are waiting on it.
   Tasks can wait on the :meth:`~Barrier.wait` method and would be blocked until
   the specified number of tasks end up waiting on :meth:`~Barrier.wait`.
   At that point all of the waiting tasks would unblock simultaneously.

   The barrier can be reused any number of times.

   .. _asyncio_example_barrier:

   Example::

      async def example_barrier():
         # action when barrier will unblock
         async def done():
            print("calling action 'done'")

         # barrier with 3 parties
         b = asyncio.Barrier(3, done)

         # create 2 new waiting tasks
         asyncio.create_task(b.wait())
         asyncio.create_task(b.wait())

         await asyncio.sleep(0)
         print(b)

         # Now, current task waits
         async with b as pos:
            print("barrier passed")

         await asyncio.sleep(0)
         print(b)

   Result of this example is:

      >>> asyncio.run(example_barrier())
      <asyncio.locks.Barrier object at 0x103e58bf0 [unlocked, wait:2/3, state:0]>
      calling action 'done'
      barrier passed
      <asyncio.locks.Barrier object at 0x103e58bf0 [unlocked, state:0]>

>    As shown in this example, the :meth:`~Barrier.wait` method: can also be call with
>    :keyword:`async with` statement.

Reviewer 2: ```suggestion
   The example also demonstrates using `async with` as 
   an alternative to awaiting on ``barrier.wait()``.
```

>    The barrier can be reused any number of times for the same number of tasks.

Reviewer 2: Drop this sentence, we already have it few lines earlier.

Author: Done

.. class:: Barrier(parties, action=None)

   Create a barrier object for *parties* number of tasks.  A coroutine *action*,

>    when provided, is awaited once just before the release phasis (draining state).

Reviewer 2: ```suggestion
   when provided, is awaited once just before the release occurs (draining state).
```

   .. coroutinemethod:: wait()

      Pass the barrier.  When all the tasks party to the barrier have called
      this function, they are all unblocked simultaneously.

      When a task in the barrier is cancelled, two cases are possible:

>       * the task is the last waiting task, an :class:`BrokenBarrierError`

Reviewer 2: ```suggestion
      * if the task is the last waiting task, a :class:`BrokenBarrierError`
```

>         exception occurs, barrier is put into a broken state;

Reviewer 2: ```suggestion
        exception is raised. The barrier is put into a broken state;
```

>       * the task is waiting (filling state) or blocked (draining state),

Reviewer 2: ```suggestion
      * if the task is waiting (filling state) or blocked (draining state),
```

        this task exits the barrier which stays in the same state.

>         If the state is filling, count of waiting task decreases by 1.

Reviewer 2: ```suggestion
        If the state of the barrier is "filling", the number of waiting task decreases by 1.
```

>       The return value is an integer in the range 0 to *parties-1*, different

Reviewer 2: ```suggestion
      The return value is an integer in the range of 0 to ``parties-1``, different
```

   .. versionchanged:: 3.10
      Removed the *loop* parameter.

Barrier
=======

   .. versionadded:: 3.11

   A barrier object.  Not thread-safe.

>    This class, as a clone of :class:`threading.Barrier`, provides a simple synchronization
>    primitive for use by a fixed number of tasks that need to wait for each other.
>    Each of the tasks tries to pass the barrier by calling the :meth:`~Barrier.wait` method
>    and would block until all of the tasks have made their :meth:`~Barrier.wait` calls.
>    At this point, the tasks are released simultaneously.

Reviewer 2: ```suggestion
   A barrier is a simple synchronization primitive that allows to block until a
   certain number of tasks are waiting on it.   
   Tasks can wait on the :meth:`~Barrier.wait` method and would be blocked until 
   the specified number of tasks end up waiting on :meth:`~Barrier.wait`.
   At that point all of the waiting tasks would unblock simultaneously.
```

>    The barrier can be reused any number of times for the same number of tasks.

Reviewer 2: ```suggestion
   The barrier can be reused any number of times.
```

          print(f'\tFile {filename!r} is removed')

          # wait for other files
          pos = await barrier.wait()

          return pos

       async def remove_folder(barrier, folder):
          print(f'Waiting for remove all files from folder {folder!r} ...')

          # wait for all removed files
          async with barrier as pos:
            # remove folder (simulation with 1 sec sleeping)
            await asyncio.sleep(1.0)
            print(f'Folder {folder!r} is now removed ...')

            return pos

       async def main(folder, files, nb_files):
          # Create a Barrier object with parties equal to nb_files+1
          # corresponding to the number of all removing items plus the folder,
          # and an action as a print message.
          barrier = asyncio.Barrier(nb_files+1, action=job_done)

          # Add remove_file task for each filename.
          tasks = [asyncio.create_task(remove_file(barrier, f)) for f in files]

          # Add remove_folder task for the folder.
          tasks.append(asyncio.create_task(remove_folder(barrier, folder)))

          # Wait until folder+filenames are removed.
          await asyncio.gather(*tasks)

>        asyncio.run(main("MyFolder", [f'file{i+1:02d}.txt' for i in range(NB)], NB))

Reviewer 2: This example is rather long and contrived. I'd drop it.

Author: Yes I agree. Do we need a simplier example ?

Author: I proposed an another example

   In the previous example, there are the two ways to call the :meth:`~Barrier.wait` method:

   * with a **classic call**: see ``pos = barrier.wait()`` in the 'remove_file' coroutine;

   * with a **async context manager call**: see ``async with barrier as pos:`` in the 'remove_folder' coroutine.

   The barrier can be reused any number of times for the same number of tasks.

.. class:: Barrier(parties, action=None)

   Create a barrier object for *parties* number of tasks.  A coroutine *action*, when
   provided, is awaited once just before the release phasis (draining state).

   .. coroutinemethod:: wait()

      Pass the barrier.  When all the tasks party to the barrier have called

>       this function, they are all released simultaneously.

Reviewer 2: ```suggestion
      this function, they are all unblocked simultaneously.
```

      When a task in the barrier is cancelled, two cases are possible:

      * the task is the last waiting task, an :class:`BrokenBarrierError`
        exception occurs, barrier is put into a broken state;

      * the task is waiting (filling state) or blocked (draining state),
        this task exits the barrier which stays in the same state.
        If the state is filling, count of waiting task decreases by 1.

      The return value is an integer in the range 0 to *parties-1*, different
      for each task. This can be used to select a task to do some special
      housekeeping, e.g.::

         ...
         async with barrier as position:
            if position == 0:
               # Only one task print this
               print('End of *draining phasis*')

      If an *action* was provided as a coroutine to the constructor, the last calling task of
      :meth:`wait` method will have to *await this coroutine* prior to being released.

>       Should this call raise an error, including cancellation of task, the barrier is put into the broken state.

Reviewer 2: Long line, please trim to 79

Author: Done

      :ref:`What's New in 3.10's Removed section <whatsnew310-removed>`
      for more information.

Barrier
=======

>    .. versionadded:: 3.10

Reviewer 2: It's too late to push this to 3.10, but 3.11 is OK.

Author: Ok, I will change

   A barrier object.  Not thread-safe.

   This class, as a clone of :class:`threading.Barrier`, provides a simple synchronization
   primitive for use by a fixed number of tasks that need to wait for each other.

>    Each of the task tries to pass the barrier by calling the :meth:`~Barrier.wait` method

Reviewer 2: ```suggestion
   Each of the tasks tries to pass the barrier by calling the :meth:`~Barrier.wait` method
```

>    and will block until all of the tasks have made their :meth:`~Barrier.wait` calls.

Reviewer 2: ```suggestion
   and would block until all of the tasks have made their :meth:`~Barrier.wait` calls.
```

          print(f'Waiting for remove all files from folder {folder!r} ...')

          # wait for all removed files
          await barrier.wait()

          # remove folder (simulation)
          await asyncio.sleep(1.0)
          print(f'Folder {folder!r} is now removed ...')

          return -1

       async def main(folder, files, nb_files):
          # Create a Barrier object with parties equal to nb_files+1
          # corresponding to the number of all removing items,
          # and an action as a print message.
          barrier = asyncio.Barrier(nb_files+1, action=job_done)

          # Add remove_file task for each filename.
          tasks = [asyncio.create_task(remove_file(barrier, f)) for f in files]

          # Add remove_folder task for the folder.
          tasks.append(asyncio.create_task(remove_folder(barrier, folder)))

          # Wait until folder+filenames are removed.
          await asyncio.gather(*tasks)

       asyncio.run(main("Datas", [f'file{i+1:02d}.txt' for i in range(NB)], NB))

.. class:: Barrier(parties, action=None)

   Create a barrier object for *parties* number of tasks.  A callable *action*, when

>    provided, is called once just before the release phasis (draining state).

Reviewer 2: I'm wondering if the design of this primitive should deviate from the threading version. In this case, I think we should rely on `async with` instead of passing a callback (which are always a PITA to deal with in async/await context) :

```python
async with asyncio.Barrier(10) as b:
    tasks = [asyncio.create_task(remove_file(barrier, f)) for f in files]
    tasks.append(asyncio.create_task(remove_folder(barrier, folder)))
    await asyncio.gather(*tasks)

msg = 'ACTION: All files removed'
print(f'{os.linesep}{msg:-^80s}{os.linesep}')
```

Author: Thank you @1st1 for your response. 

Notice: when specified, the callback runs **once** from the last task calling the  `wait` method and just before draining the barrier. If the callback crashes, the barrier state goes to broken.

Are you suggesting to remove that callback parameter or to replace with a coro and run an `awaiting` of it ? I agree to respect the asyncio context, but in that case I don't understand a PITA with this callback. May I miss something ?

Moreother, `__aenter__` and `__aexit__`  methods  are missing in the Barrier class. To add with dedicated tests.
